File: Leetcode/7.cpp
/**
 * 7. 整数反转 处理溢出
 * link：https://leetcode-cn.com/problems/reverse-integer/
 */

class Solution
{
public:
    int reverse(int x)
    {
        int result = 0;
        constexpr int int_min_div = INT_MIN / 10;
        constexpr int int_max_div = INT_MAX / 10;
        constexpr int int_min_remain = INT_MIN % 10;
        constexpr int int_max_remain = INT_MAX % 10;

        while (x != 0)
        {
            int remain = x % 10;
            if (result > int_max_div || (result == int_max_div && remain > int_max_remain))
                return 0;
            if (result < int_min_div || (result == int_min_div && remain < int_min_remain))
                return 0;
            result = result * 10 + remain;
            x = x / 10;
        }
        return result;
    }
};

File: Leetcode/7.rb
# 7. 整数反转 处理溢出
# link：https://leetcode-cn.com/problems/reverse-integer/

# @param {Integer} x
# @return {Integer}
def reverse(x)
  result = 0
  negative = (x < 0)
  x = x.abs
  while x > 0
    result *= 10
    if result > ((1 << 31) - (negative ? 0 : 1)) - x % 10
      return 0
    end
    result += x % 10
    x /= 10
  end
  negative ? -result : result
end
